ar/projects with many files (#106)
Ada-Developers-Academy/core-unit-1: intro-to-oop-contd/projects-with-many-files.md

## Vocabulary and Synonyms

| Vocab   | Definition                                                                             | How to Use in a Sentence                                                                                                                  |
| ------- | -------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------- |
| Module  | Any `.py` file that contains functions, classes, variables, and/or other runnable code | "I used the `math.Euler()` from the `Math` module to calculate an angle"                                                                  |

> | Package | A collection of modules                                                                | "The `requests` package contains modules to help us make HTTP requests." "My package has custom configuration in its `__init__.py` file." |

Author: ```suggestion
| Package | A collection of modules                                                                | "The `requests` package contains modules to help us request data from the Internet." "My package has custom configuration in its `__init__.py` file." |
```

> | Package | A collection of modules                                                                | "The `requests` package contains modules to help us make HTTP requests." "My package has custom configuration in its `__init__.py` file." |

Author: Not motivated enough to think of a different package, but I can rephrase it to something more easy to identify with.

### Project Code Lives in a Package Folder

> Under the project root, we will have a single folder to contain all of our project classes, one class per `.py` file. We refer to this folder as the **package folder**. At the start of a project, when we don't have many class files, we might put them all directly in the package folder. But as the project becomes more complex, we can add subfolders to further group together related class files.

Author: I'll take this for now as a reminder to address this while reworking this part to get rid of the __main__ stuff. I might be able to bring the passage that used to follow this back up as part of the rearrangement, making this line specifically redundant, but it'll get addressed one way or the other.

There are also a number of special files that can be placed in the package folder.

The most common of these is called the `__init__.py` file, which is discussed briefly in the Package Details section below.

Projects with a clear starting point might include a `__main__.py` as a fairly standard place to put the application start up logic. Some projects may not need this file, and some might choose to put that logic somewhere else.

Our package folder usually has the same name as the project, but it must follow some package naming rules which we will review a little later in this lesson.

> The above example names this folder `project_package_name` so that we can distinguish it from the project root, but there would be nothing preventing us from calling it `project_name` to match the project root name. It contains two class files, `example_class_a.py` and `example_class_b.py`, and both an `__init__.py` file and a `__main__.py`.

Author: Just gonna leave this as is, since I need to cull the __main__ stuff anyway. This will get addressed as part of that.

### Importing from Packages and Modules

There are three common ways that we will see packages and modules imported into our projects.

Regardless of the `import` style used, we should prefer to put `import` statements near the top of the file doing the importing. It _is_ valid syntax to put `import` statements elsewhere in a file, but by putting them near the top, other programmers will know where to look for the modules on which our module depends.

#### 1. Importing a Module by Name

This method is especially common for modules that are built-in to Python. We will tend _not_ to use this method when importing modules from our own packages, but we should be comfortable seeing it in other code, and using it to access built-in Python features.

As an example, consider the `random` module, which provides the `randint` function.

```python
import random
```

> This statement imports the `random` module name into the current scope, allowing us to access identifiers in the module by using dot notation to look inside the module name as follows:

Author: 'This statement' refers back to the previous statement. I'd still like a for example leading into the next.

```python
import ride_share_app.driver
```

This statement imports the `ride_share_app` package name into the current scope. It further imports the `driver` module as a member of the `ride_share_app` package, allowing us to access identifiers in the module by using dot notation. Consider:

```python
driver = ride_share_app.driver.Driver()
```

Assuming there is an identifier called `Driver` in the `driver` module and that it is a class, this would make a new instance and store a reference to it in the `driver` variable. Notice that there is no issue with having a variable in the current scope that matches a module name in a package. The `driver` variable name and the `driver` module name exist in two separate scopes.

### !callout-info

## Why We Keep Saying Identifier

Notice that we have been careful to say that Python imports identifiers into the current scope. Identifiers are names of things. Variables, functions, and classes are all named with identifiers. From just looking at an identifier, we don't know what it refers to. So we have to be careful to import the correct identifiers, and then use them as they are intended to be used.

<br />

> Hopefully good naming practices were employed to help hint at the proper usage of different identifiers. But if we import an identifier that refers to a value (a variable) but then try to call it like a function, this will result in a runtime error! So it's our job to read and understand the code or documentation of the module we want to use so that we import and make use of its identifiers correctly!

Author: I really wish we had a part two to this lesson, but I'm sure that won't happen just now. Not necessarily here, but later in the curriculum. It feels like we're just scratching the surface in so many places, and I worry that we haven't built enough of a foundation in this one lesson (which is fairly disconnected from much of the rest of the material) for students to evaluate additional information sources. I think that contributes to my desire to talk about that next little bit, and the next little bit!

### !end-callout

#### 3. Importing Module Identifiers using Relative Names

Importing package modules by full name works everywhere, whether from outside or inside a package. But to access the identifiers defined in the module, we have to prefix them with the full package name. If our package were several layers deep, this could make for some long names!

> Let's say that the `passenger` module in our `ride_share_app` package needs to make use of the `Driver` class. We could certainly import the `driver` module as before and access the `Driver` class with the full package name. But `driver.py` and `passenger.py` are defined in the same package. In fact, they're in the same folder. Wouldn't it be nice if there were a shorter way?

Reviewer: I know it's a lot of copy/pasting, but I think that copy/pasting and repeating the `ride_share_app` structure before this paragraph would help visualize to students that `driver.py` and `passenger.py` are in the same folder.

There is!

```python
from .driver import Driver
```

This may look a little weird. It might even seem backwards! Feel free to read this statement from right to left if that's helpful!

What this statement does is:

1. Use the `from` keyword to name the module from which we want to import an identifier.
1. Use the `import` keyword to name the identifier we want Python to import into the current scope

The module name looks different from the examples we've seen before. It starts with a `.` which tells Python to look for the module in the folder containing the current file. When Python finds the specified module, rather than importing the name of the module into the current scope, it looks at the names listed after `import` (we can write multiple names separated by commas) and imports just those identifiers into the current scope.

If we assume as before that the `Driver` identifier is a class, this lets us write the following:

```python
driver = Driver()
```

Shorter and clearer!

```
└── ride_share_app
    ├── __init__.py
    ├── driver.py
    ├── passenger.py
    └── payment
        ├── __init__.py
        ├── cash.py

>         └── credit.py

Author: WHAT IS THIS WITCHCRAFT?!?!?!?

I did these all MANUALLY!!!!!

It was really that bad. I just copied an existing tree layout and used area selections to move bits around.... BUT STILL!!!!!

In any case, what happened there was that I didn't want to bring down all the test directory stuff, so initially I tried inserting some ellipses, but I didn't like it. So then I just had a dangly down bit that trailed into nothing, and I didn't like that. So then I trimmed it as you see. But I guess that's weird too.

Long story short, I'll take your suggestion, then sit in a corner and sob softly.

```

> If our package had a folder called `payment` under `ride_share_app` (with an `__init__.py` file), and `payment` had the modules `cash.py` and `credit.py`, then from either of these modules we could access the `Driver` identifier in `driver.py` as:

Reviewer: My preference is to think of this `payment` section as a second paragraph.

With that, I think I'd recommend:

- Moving this line to directly above the example tree structure (currently directly above this paragraph)
- Modifying the wording of this line to introduce this as a second example
- Frame this second example to solving a different question: What's the syntax for importing something from two folders up?
- Name the modules `cash.py` and `credit.py` in a simplified sense instead of saying "then from either of these modules." (take a more concrete stance, so a student brain doesn't have to parse through "is it for cash.py or credit.py, what is a module, which folder are we looking at," etc)

```suggestion
Let's consider a second, similar situation. How do we import a module that is inside a different folder, but within the same package?

Imagine that our package had a folder called `payment` under `ride_share_app` (with an `__init__.py` file), and `payment` had the modules `cash.py` and `credit.py`. Inside either `cash.py` or `credit.py`, we can access the `Driver` identifier in `driver.py` as:
```

Author: Taking the suggestion. I'll move the lines around locally.

```python
from module_one import UsefulClass as UsefulClassOne
from module_two import UsefulClass as UsefulClassTwo
```

Instead of importing the identifier `UsefulClass` into the current scope and then overwriting it, this second version imports each of the `UsefulClass` variables into two separate identifiers: `UsefulClassOne`, and `UsefulClassTwo`, which we can use as:

```python
one = UsefulClassOne()
one.do_something()
two = UsefulClassTwo()
two.do_something_else()
```

### So Many Ways to Import

As we have shown, there are _many_ different syntaxes and strategies we can use to import modules. If a particular import syntax works for a project, it's valid!

Different situations will need to use different ways of importing. So we should be ready to see many styles of importing. We may even encounter styles that we haven't discussed that we might need to look up. That's ok! But we should feel confident that we can recognize importing when we see it!

> ## Restructuring a Project

Reviewer: Soft suggestion/question? I found myself craving that this section was framed as an example walkthrough. Your first sentence in this section pretty much states that though. Would it be valuable to rename this to something like:

```suggestion
## Example: Restructuring a Project
```

```suggestion
## Example Walkthrough: Restructuring a Project
```

or something to that effect?

Let's return to Scarlet's project and see how she decides to reorganize her project. Remember that she started with a single file `main.py` that contained two classes, `Driver` and `Passenger`, and there was some main logic to create a `Passenger` instance.

Scarlet is making a ride share app, so she decides to call her project root `ride-share-app`. She puts her standard project files in the project root. She knows that in Python we like to have each class in its own module, and that the modules for a project are usually grouped into a package. She can't use hyphens in the package name, so she decides on `ride_share_app` as her package name. She makes the folder, adds the `__init__.py` file to tell Python this is a package folder, then makes a module file for each class: `driver.py` and `passenger.py`.

> She considers renaming the `main.py` which now only contains the application startup to `ride_share.py` to be a little more descriptive on the command line. But Scarlet learned about another strategy for where to put her startup code that she wants to try, so for now she leaves the code in `main.py`
> 
> ### !callout-warning
> 
> ## Be Careful With Descriptive Entry Point File Names
> 
> Scarlet also briefly considered renaming `main.py` to `ride_share_app.py`, but then she realized that would prevent Python from being able to see the `ride_share_app` package. If the main file has the same name as the package, if we try to import from the package, Python will try looking inside the module we are currently in, not find the desired identifier, and raise an `ImportError`.
> 
> <br />
> 
> We should make sure that the entry point file name does _not_ match the project package name. Alternatives include adding `_pkg` to the name of the package, though this isn't very Pythonic. Or we could make the entry point closely related to, but not the same as the package name, like how Scarlet considered `ride_share.py`. We can use a more general entry point, like `main.py`. Or if we want to be able to use the package name to launch the application, we can use the special `__main__.py` file described later in this lesson.
> 
> ### !end-callout

Reviewer: The students don't necessarily have meaningful context or examples to have run into this, felt this pain, or the motivation to do something with this knowledge, IMO.

I think it personally takes until capstone to begin to have opinions on file names.

Also, the concept of an "entry point" is pretty intangible and far to them IMO. The closest entry point that they may see may be a Flask project depending on how you think about it. Even at that point, I think it's a lot of explanation to think about an entry point and it'd get awkward without the proper way to visualize it.

In that whole theme, I think that the idea of Scarlet considering renaming `main.py` is a really out-of-reach consideration. I don't think the knowledge/scenario presented here will be the juiciest for students.

Personally, I'm into the idea of cutting it, to make room for the other good content in here?

```suggestion
```

Author: You already heard the back story for this. But as we're going to deemphasize the entry point stuff, this can go!

We have already seen the special `__init__.py` file that Python uses to mark package folders. There is another special package file that can help us organize the way we start our programs.

<br />

If our package includes a file called `__main__.py`, we can start our programs like `python3 -m package_name`. We won't go into the details here, but follow your curiosity!

### !end-callout

### Don't Forget the Tests

Scarlet hasn't written tests for her app yet. Take a moment to think about where she should add tests to this project, then reveal a possible structure below.

<details style="max-width: 700px; margin: auto;">
    <summary>
      Click to reveal a possible project layout that includes tests.
    </summary>

```
ride-share-app/
├── README.md
├── requirements.txt
├── main.py
├── ride_share_app
│   ├── __init__.py
│   ├── driver.py
│   └── passenger.py
└── tests
    ├── __init__.py
    ├── driver_test.py
    └── passenger_test.py
```

Notice that the `tests` folder is in the project root folder (`ride-share-app`), not in the package folder (`ride_share_app`).

</details>

## Debugging Imports

Here are some common bugs that we may come across with importing:

> ### `ModuleNotFoundError`
> 

Reviewer: I could imagine a small lead-up sentence:

```suggestion
### `ModuleNotFoundError`

Consider this stack trace caused from running `passenger.py`

```

Author: Ha ha.  Yeah. This struck me as a little abrupt too, just not enough to do anything about.

Thanks for the suggestions though!  I'll take them. 🙂

│       ├── __init__.py
│       ├── back_camera.py
│       └── front_camera.py
└── tests
    ├── __init__.py
    ├── processor_test.py
    ├── filters
    │   ├── __init__.py
    │   ├── cat_ears_filter_test.py
    │   └── sepia_filter_test.py
    └── image_source
        ├── __init__.py
        ├── back_camera_test.py
        └── front_camera_test.py
```

##### !end-answer

##### !explanation

A good default layout is to have our general project files in the project root. Then we create two top-level packages, one for our application code, and one for the tests. The structures under each should generally mirror one another.  Each package folder should have an `__init__.py` file. And when we start dealing with a number of files, some of which are more closely related to each other than the rest of the package, we should consider creating subfolders to group the related files.

##### !end-explanation

### !end-challenge
<!-- prettier-ignore-end -->
<!-- ======================= END CHALLENGE ======================= -->

<!-- prettier-ignore-start -->
### !challenge

* type: multiple-choice
* id: f032adcb-c05b-4591-96b9-58bc24d7ee79
* title: Projects With Many Files

##### !question

> Hosts for AdaBnB can host guests at multiple properties on the app. Each property is an instance of a `Place` class. If we tried to call the `create_places` method of `Host`, we would receive a `NameError`. Assuming the `place` module in which the `Place` class is defined is in the same package folder as `host.py`, select the option that best describes how to solve this `NameError`.

Reviewer: Soft suggestion: The sentence is good, but it might be nice to give them a diagram either additonally, or instead.

For convenience

```
adabnb/
├── README.md
├── requirements.txt
├── adabnb
│   ├── __init__.py
│   ├── host.py
│   └── place.py
└── tests
    ├── __init__.py
    ├── host_test.py
    └── place_test.py
```

Didn't make a suggestion _because code formatting_.

Author: Good idea! And thanks for the tree!

from non_existing_module import Driver
```

The `ModuleNotFoundError` is raised because the program can't find `non_existing_module`. In general, this error tells us:

- There's an `import` statement that can't find the specified module
  - Double-check the `from` part. Do we need to including the package name, if this file isn't in the same package? Ar we missing a `.` (one or more) for a relative import?

> ### `ImportError`
> 

Reviewer: Similar to above

```suggestion
### `ImportError`

Consider this stack trace caused from running `passenger.py`

```
